### PR TITLE
Orb tracking

### PR DESCRIPTION
**Pull request by misternebula**

### CODE REVIEW

File: QSB/TransformSync/NomaiOrbTransformSync.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                transform.rotation = OrbParent.InverseTransformRotation(OrbTransform.rotation);
+                return;
+            }
+            OrbTransform.position = OrbParent.TransformPoint(transform.position);
+            OrbTransform.rotation = OrbParent.InverseTransformRotation(OrbTransform.rotation);
+            if (transform.localPosition == Vector3.zero)
+            {
+                _updateCount++;
+            }
+            if (_updateCount >= 5)
```
Comment (amazingalek) [inline QSB/TransformSync/NomaiOrbTransformSync.cs:79]: what's 5?

File: QSB/OrbSync/QSBOrbSlot.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            if (!_initialized)
+            {
+                return;
+            }
+            if (state)
+            {
+                WorldRegistry.RaiseEvent(InterfaceSlot, "OnSlotActivated");
+                return;
+            }
+            WorldRegistry.RaiseEvent(InterfaceSlot, "OnSlotDeactivated");
```
Comment (amazingalek) [inline QSB/OrbSync/QSBOrbSlot.cs:38]: var ev = state ? "OnSlotActivated" : "OnSlotDeactivated";
WorldRegistry.RaiseEvent(InterfaceSlot, ev);

File: QSB/TransformSync/NomaiOrbTransformSync.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+using UnityEngine.Networking;
+
+namespace QSB.TransformSync
+{
+    public class NomaiOrbTransformSync : NetworkBehaviour
+    {
+        public NomaiInterfaceOrb AttachedOrb { get; private set; }
+        public Transform OrbTransform { get; private set; }
+
+        private int Index => WorldRegistry.OrbSyncList.FindIndex(x => x == this);
```
Comment (amazingalek) [inline QSB/TransformSync/NomaiOrbTransformSync.cs:12]: `WorldRegistry.OrbSyncList.IndexOf(this);`

File: QSB/TransformSync/NomaiOrbTransformSync.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
+﻿using QSB.WorldSync;
+using UnityEngine;
+using UnityEngine.Networking;
+
+namespace QSB.TransformSync
+{
+    public class NomaiOrbTransformSync : NetworkBehaviour
```
Comment (amazingalek) [inline QSB/TransformSync/NomaiOrbTransformSync.cs:7]: why not inherit from TransformSync?

File: QSB/TransformSync/NomaiOrbTransformSync.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        private bool _isInitialized;
+        private bool _isReady;
+        private Transform _orbParent;
+        private int _updateCount;
+
+        public override void OnStartClient()
+        {
+            WorldRegistry.OrbSyncList.Add(this);
+
+            QSB.Helper.Events.Unity.RunWhen(() => WorldRegistry.OldOrbList.Count != 0, OnReady);
```
Comment (amazingalek) [inline QSB/TransformSync/NomaiOrbTransformSync.cs:25]: `> 0`

File: QSB/TransformSync/NomaiOrbTransformSync.cs (hunk 4 above)
Comment (misternebula) [reply QSB/TransformSync/NomaiOrbTransformSync.cs:7]: don't need the sector sync

Comment (misternebula) [reply QSB/TransformSync/NomaiOrbTransformSync.cs:7]: TransformSync's variables and whole structure is based on sectors... for something to use that when it doesnt use sectors seems wrong

File: QSB/OrbSync/OrbPatches.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                WorldRegistry.HandleSlotStateChange(__instance, orb, false);
+                ____occupyingOrb = null;
+                WorldRegistry.RaiseEvent(__instance, "OnSlotDeactivated");
+                __result = false;
+                return false;
+            }
+            __result = true;
+            return false;
+        }
```
Comment (misternebula) [inline QSB/OrbSync/OrbPatches.cs:53]: This is direct imported OW code, could probably be easily refactored

File: QSB/OrbSync/OrbSlotManager.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        private void Awake()
+        {
+            QSBSceneManager.OnSceneLoaded += OnSceneLoaded;
+        }
+
+        private void OnSceneLoaded(OWScene scene, bool isInUniverse)
+        {
+            QSB.Helper.Events.Unity.RunWhen(() => QSB.HasWokenUp, InitSlots);
+        }
```
Comment (misternebula) [inline QSB/OrbSync/OrbSlotManager.cs:16]: Could be combined?

File: QSB/OrbSync/OrbSlotManager.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+        private void InitSlots()
+        {
+            var orbSlots = Resources.FindObjectsOfTypeAll<NomaiInterfaceSlot>();
+            for (var id = 0; id < orbSlots.Length; id++)
+            {
+                var qsbOrbSlot = WorldRegistry.GetObject<QSBOrbSlot>(id) ?? new QSBOrbSlot();
+                qsbOrbSlot.Init(orbSlots[id], id);
+                WorldRegistry.AddObject(qsbOrbSlot);
```
Comment (misternebula) [inline QSB/OrbSync/OrbSlotManager.cs:25]: Could be moved to the awake of the orb slot?

File: QSB/QSBNetworkManager.cs [hunk 9]
```diff
@@ -67,8 +88,11 @@ private void ConfigureNetworkManager()
             customConfig = true;
             connectionConfig.AddChannel(QosType.Reliable);
             connectionConfig.AddChannel(QosType.Unreliable);
+            this.SetValue("m_MaxBufferedPackets", MaxBufferedPackets);
```
Comment (misternebula) [inline QSB/QSBNetworkManager.cs:91]: Probably not needed? But with the amount of network messages we send, might be a good idea

File: QSB/Utility/DebugLog.cs [hunk 10]
```diff
@@ -9,7 +9,20 @@ public static class DebugLog
...
+            var callingType = GetCallingType(new StackTrace());
+            if (callingType == "DebugLog")
+            {
+                callingType = GetCallingType(new StackTrace(), 2);
+            }
+            if (callingType == "DebugLog")
+            {
+                callingType = GetCallingType(new StackTrace(), 3);
+            }
+            method.Invoke(console, new object[] { type, message, callingType });
```
Comment (misternebula) [inline QSB/Utility/DebugLog.cs:25]: Sorry alek 😅

File: QSB/Utility/DebugLog.cs [hunk 11]
```diff
@@ -1,12 +1,28 @@
...
+                .Last(x => x.Name == "WriteLine");
+            var callingType = GetCallingType(new StackTrace());
+            if (callingType == "DebugLog")
+            {
+                callingType = GetCallingType(new StackTrace(), 2);
+            }
+            if (callingType == "DebugLog")
+            {
+                callingType = GetCallingType(new StackTrace(), 3);
+            }
```
Comment (amazingalek) [inline QSB/Utility/DebugLog.cs:24]: can you build this into GetCallingType somehow?

File: QSB/Utility/DebugLog.cs (hunk 10 above)
Comment (amazingalek) [reply QSB/Utility/DebugLog.cs:25]: Seems like OWML should support another overload for WriteLine?

Comment (misternebula) [reply QSB/Utility/DebugLog.cs:25]: I suggested that before, but you didn't seem too keen on mods being able to spoof log location easily

**Review by amazingalek — DISMISSED** [review]
👍👍👍👍👍

**Pull request merged**